src/librbd/Operations.cc

```diff
@@ -1024,6 +1025,46 @@ void Operations<I>::execute_snap_unprotect(const char *snap_name,
 }
 
 template <typename I>
+int Operations<I>::set_snap_limit(const int limit) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 5) << this << " " << __func__ << ": limit=" << limit << dendl;
+
+  if (m_image_ctx.read_only) {
+    return -EROFS;
+  }
+
+  int r = m_image_ctx.state->refresh_if_required();
+  if (r < 0) {
+    return r;
+  }
+
+  {
+    RWLock::RLocker owner_lock(m_image_ctx.owner_lock);
+    C_SaferCond cond_ctx;
+    execute_snap_limit(limit, &cond_ctx);
+
+    r = cond_ctx.wait();
+  }
+
+  return r;
+}
+
+template <typename I>
+void Operations<I>::execute_snap_limit(const uint64_t limit,
```

Comment:
Probably overly complex -- I don't think there is a need to proxy this request.  The rationale is that the limit won't affect the object map, journaling, etc (things that the exclusive lock is protecting).  I think you can just create a small, synchronous helper method to invoke the "snapshot_set_limit" cls_client helper method.